### PR TITLE
PR1 for AlicaEngine refactory

### PR DESCRIPTION
**Pull request by triccyx**
First PR for AlicaEngine refactory 
ModelManager and subclasses

### CODE REVIEW

File: alica_engine/include/engine/model/AbstractPlan.h [hunk 1]
```diff
@@ -25,8 +25,12 @@ class AlicaEngine;
...
 {
 public:
-    AbstractPlan(AlicaEngine* ae);
-    AbstractPlan(AlicaEngine* ae, int64_t id);
+    //[[deprecated("It will be removed in the last PR")]]
+    AbstractPlan(AlicaEngine* ae); // TOBE removed
+    //[[deprecated("It will be removed in the last PR")]]
+    AbstractPlan(AlicaEngine* ae, int64_t id); // TOBE removed
+    AbstractPlan(const YAML::Node& config);
+    AbstractPlan(const YAML::Node& config, int64_t id);
```
Comment (abhi1287) [inline alica_engine/include/engine/model/AbstractPlan.h]: config is unused now.

File: alica_engine/include/engine/model/Behaviour.h [hunk 2]
```diff
@@ -28,7 +28,9 @@ class AlicaEngine;
 class Behaviour : public AbstractPlan
 {
 public:
-    Behaviour(AlicaEngine* ae);
+    //[[deprecated("It will be removed in the last PR")]]
+    Behaviour(AlicaEngine* ae); // TOBE removed
+    Behaviour(const YAML::Node& config);
```
Comment (abhi1287) [inline alica_engine/include/engine/model/Behaviour.h]: unused config

**Comment by triccyx** [comment]
@abhi1287

File: alica_engine/include/engine/model/Plan.h [hunk 3]
```diff
@@ -19,13 +19,16 @@ class PreCondition;
...
 /**
  * An ALICA plan
  */
 class Plan : public AbstractPlan
 {
 public:
-    Plan(AlicaEngine* ae, int64_t id);
+    //[[deprecated("It will be removed in the last PR")]]
+    Plan(AlicaEngine* ae, int64_t id); // TOBE removed
+    Plan(const YAML::Node& config, ConfigChangeListener& configChangeListener, int64_t id);
```
Comment (abhi1287) [inline alica_engine/include/engine/model/Plan.h]: to further simplify, config could be part of ConfigChangeListener and could be exposed via a getConfig() method.

File: alica_engine/include/engine/model/Plan.h [hunk 4]
```diff
@@ -56,6 +59,11 @@ class Plan : public AbstractPlan
 
     std::string toString(std::string indent = "") const;
 
+    AlicaTime getAuthorityTimeInterval() const { return _authorityTimeInterval; }
+    void setAuthorityTimeInterval(AlicaTime authorityTimeInterval) const; // not a mistake, this is mutable
+
+    void reload(const YAML::Node& config);
```
Comment (abhi1287) [inline alica_engine/include/engine/model/Plan.h:65]: could be private

File: alica_engine/include/engine/model/PlanType.h [hunk 5]
```diff
@@ -18,7 +18,9 @@ class AlicaEngine;
 class PlanType : public AbstractPlan
 {
 public:
-    PlanType(AlicaEngine* ae);
+    //[[deprecated("It will be removed in the last PR")]]
+    PlanType(AlicaEngine* ae); // TOBE removed
+    PlanType(const YAML::Node& config);
```
Comment (abhi1287) [inline alica_engine/include/engine/model/PlanType.h]: unused config

File: alica_engine/include/engine/modelmanagement/factories/BehaviourFactory.h [hunk 6]
```diff
@@ -10,7 +10,10 @@ class Behaviour;
 class BehaviourFactory : public Factory
 {
 public:
-    static Behaviour* create(AlicaEngine* ae, const YAML::Node& node);
+    //[[deprecated("It will be removed in the last PR")]]
+    static Behaviour* create(AlicaEngine* ae, const YAML::Node& node); // TOBE removed
+    static Behaviour* create(const YAML::Node& config, const YAML::Node& node);
```
Comment (abhi1287) [inline alica_engine/include/engine/modelmanagement/factories/BehaviourFactory.h]: unused config

File: alica_engine/include/engine/modelmanagement/factories/PlanTypeFactory.h [hunk 7]
```diff
@@ -10,7 +10,9 @@ class PlanType;
 class PlanTypeFactory : public Factory
 {
 public:
-    static PlanType* create(AlicaEngine* ae, const YAML::Node& planTypeNode);
+    //[[deprecated("It will be removed in the last PR")]]
+    static PlanType* create(AlicaEngine* ae, const YAML::Node& planTypeNode); // TOBE removed
+    static PlanType* create(const YAML::Node& config, const YAML::Node& planTypeNode);
```
Comment (abhi1287) [inline alica_engine/include/engine/modelmanagement/factories/PlanTypeFactory.h]: unused config

File: alica_engine/src/engine/AlicaEngine.cpp [hunk 8]
```diff
@@ -192,7 +192,7 @@ IAlicaWorldModel* AlicaEngine::getWorldModel() const
 
 void AlicaEngine::subscribe(std::function<void(const YAML::Node& config)> reloadFunction)
 {
-    _configChangeListenerCBs.push_back(reloadFunction);
+    _configChangeListener.subscribe(reloadFunction);
```
Comment (abhi1287) [inline alica_engine/src/engine/AlicaEngine.cpp:196]: i suppose AlicaEngine::subscribe would be deprecated?

File: alica_engine/include/engine/Types.h [hunk 9]
```diff
@@ -55,6 +57,9 @@ using VariableGrp = std::vector<const Variable*>;
 using ParameterMap = std::unordered_map<std::string, Parameter*>;
 using AgentStatePair = std::pair<AgentId, const State*>;
 
+using ReloadFunction = std::function<void(const YAML::Node&)>;
+using ConfigChangeSubscriber = std::function<void(ReloadFunction)>;
```
Comment (abhi1287) [inline alica_engine/include/engine/Types.h]: could be part of ConfigChangeListener.h

File: alica_engine/include/engine/model/Plan.h (hunk 3 above)
Comment (triccyx) [reply alica_engine/include/engine/model/Plan.h]: Done

File: alica_engine/include/engine/model/AbstractPlan.h (hunk 1 above)
Comment (triccyx) [reply alica_engine/include/engine/model/AbstractPlan.h]: ok

File: alica_engine/include/engine/model/Plan.h (hunk 4 above)
Comment (triccyx) [reply alica_engine/include/engine/model/Plan.h:65]: ok

File: alica_engine/include/engine/model/PlanType.h (hunk 5 above)
Comment (triccyx) [reply alica_engine/include/engine/model/PlanType.h]: ok

File: alica_engine/include/engine/modelmanagement/factories/BehaviourFactory.h (hunk 6 above)
Comment (triccyx) [reply alica_engine/include/engine/modelmanagement/factories/BehaviourFactory.h]: ok

File: alica_engine/include/engine/modelmanagement/factories/PlanTypeFactory.h (hunk 7 above)
Comment (triccyx) [reply alica_engine/include/engine/modelmanagement/factories/PlanTypeFactory.h]: ok

File: alica_engine/src/engine/AlicaEngine.cpp (hunk 8 above)
Comment (triccyx) [reply alica_engine/src/engine/AlicaEngine.cpp:196]: Yes comment added

File: alica_engine/include/engine/Types.h (hunk 9 above)
Comment (triccyx) [reply alica_engine/include/engine/Types.h]: ok

File: alica_engine/src/engine/modelmanagement/ModelManager.cpp [hunk 10]
```diff
@@ -28,27 +29,41 @@ namespace alica
...
+    auto reloadFunctionPtr = std::bind(&ModelManager::reload, this, std::placeholders::_1);
+    _configChangeListener.subscribe(reloadFunctionPtr);
+    reload(_configChangeListener.getConfig());
     Factory::setModelManager(this);
 }
 
 void ModelManager::reload(const YAML::Node& config)
 {
+    YAML::Node& originalConfig = _configChangeListener.getConfig();
+    originalConfig = config;
```
Comment (abhi1287) [inline alica_engine/src/engine/modelmanagement/ModelManager.cpp]: What's the point of doing this? Also incoming argument of config could be further passed to getBasePath.

**Review by abhi1287 — APPROVED** [review]

File: alica_engine/src/engine/modelmanagement/ModelManager.cpp (hunk 10 above)
Comment (triccyx) [reply alica_engine/src/engine/modelmanagement/ModelManager.cpp]: Just something that is forgotten during a test

**Pull request merged**